File: packages/protocols/x402-ampersend-sdk/python/ampersend-sdk/src/ampersend_sdk/x402/treasurer.py
from abc import ABC, abstractmethod
from typing import Any, NamedTuple

from x402.types import PaymentRequirements
from x402_a2a import (
    PaymentPayload,
    PaymentStatus,
    x402PaymentRequiredResponse,
)
from x402_a2a.types import PaymentStatus, x402PaymentRequiredResponse


class X402Authorization(NamedTuple):
    """Result of payment authorization containing payment details and ID."""

    payment: PaymentPayload
    authorization_id: str
    selected_requirement: PaymentRequirements


class X402Treasurer(ABC):
    @abstractmethod
    async def onPaymentRequired(
        self,
        payment_required: x402PaymentRequiredResponse,
        context: dict[str, Any] | None = None,
    ) -> X402Authorization | None:
        """Authorize or reject a payment."""

    @abstractmethod
    async def onStatus(
        self,
        status: PaymentStatus,
        authorization: X402Authorization,
        context: dict[str, Any] | None = None,
    ) -> None:
        """Handle payment status updates."""

File: packages/protocols/x402-ampersend-sdk/typescript/packages/ampersend-sdk/src/x402/treasurer.ts
import type { PaymentPayload, PaymentRequirements } from "x402/types"

/**
 * Context information for payment decisions
 */
export interface PaymentContext {
  method: string
  params: any
  metadata?: Record<string, unknown>
}

/**
 * Authorization linking a payment with a tracking ID
 */
export interface Authorization {
  payment: PaymentPayload
  authorizationId: string
}

/**
 * Payment status types for tracking payment lifecycle
 */
export type PaymentStatus =
  | "sending" // Payment submitted to seller
  | "accepted" // Payment verified and accepted
  | "rejected" // Payment rejected by seller
  | "declined" // Buyer declined to pay
  | "error" // Error during payment processing

/**
 * X402Treasurer interface - Separates payment decision logic from payment creation
 *
 * An X402Treasurer decides whether to approve or reject payment requests,
 * tracks payment status, and delegates actual payment creation to an X402Wallet.
 *
 * @example
 * ```typescript
 * class BudgetTreasurer implements X402Treasurer {
 *   constructor(private wallet: X402Wallet, private dailyLimit: number) {}
 *
 *   async onPaymentRequired(requirements, context) {
 *     if (this.wouldExceedBudget(requirements[0])) {
 *       return null // Decline
 *     }
 *     const payment = await this.wallet.createPayment(requirements[0])
 *     return { payment, authorizationId: crypto.randomUUID() }
 *   }
 *
 *   async onStatus(status, authorization, context) {
 *     console.log(`Payment ${authorization.authorizationId}: ${status}`)
 *   }
 * }
 * ```
 */
export interface X402Treasurer {
  /**
   * Called when payment is required.
   *
   * @param requirements - Array of payment requirements from seller (typically use first)
   * @param context - Optional context about the request requiring payment
   * @returns Authorization to proceed with payment, or null to decline
   */
  onPaymentRequired(
    requirements: ReadonlyArray<PaymentRequirements>,
    context?: PaymentContext,
  ): Promise<Authorization | null>

  /**
   * Called with payment status updates throughout the payment lifecycle.
   *
   * @param status - Current payment status
   * @param authorization - The authorization returned from onPaymentRequired
   * @param context - Optional context about the status update
   */
  onStatus(status: PaymentStatus, authorization: Authorization, context?: PaymentContext): Promise<void>
}
